gh-badges/lib/make-badge.js

```diff
@@ -108,75 +25,47 @@ module.exports = function makeBadge({
   // String coercion and whitespace removal.
   text = text.map(value => `${value}`.trim())
 
-  let [left, right] = text
+  const [label, message] = text
 
-  color = normalizeColor(color)
-  labelColor = normalizeColor(labelColor)
+  color = normalizeColor(color || colorB || colorscheme)
+  labelColor = normalizeColor(labelColor || colorA)
 
   // This ought to be the responsibility of the server, not `makeBadge`.
   if (format === 'json') {
     return JSON.stringify({
-      label: left,
-      message: right,
+      label,
+      message,
       logoWidth,
       color,
       labelColor,
       link: links,
-      name: left,
-      value: right,
+      name: label,
+      value: message,
     })
   }
 
-  if (!(template in templates)) {
-    if (template === 'popout-square') {
-      template = 'flat-square'
-    } else {
-      template = 'flat'
-    }
-  }
-  if (template === 'social') {
-    left = capitalize(left)
-  } else if (template === 'for-the-badge') {
-    left = left.toUpperCase()
-    right = right.toUpperCase()
+  template = camelcase(template)
+  if (template.startsWith('popout')) {
+    template = template.replace('popout', 'flat')
   }
 
-  let leftWidth = (anafanafo(left) / 10) | 0
-  // Increase chances of pixel grid alignment.
-  if (leftWidth % 2 === 0) {
-    leftWidth++
-  }
-  let rightWidth = (anafanafo(right) / 10) | 0
-  // Increase chances of pixel grid alignment.
-  if (rightWidth % 2 === 0) {
-    rightWidth++
+  let render = badgeRenderers[template]
+  if (render === undefined) {
+    render = badgeRenderers.flat
```

Comment:
As it stands, if you're a library user template is currently optional: https://github.com/badges/shields/blob/c05394d023e89334eaa46b38fd8c276a094d62a0/gh-badges/lib/index.js#L73
Are you saying we should make template a required param in the library, or do you just mean we should keep the existing behaviour but push the responsibility for allowing this to be optional out a layer in the application?

If we were going to get rid of the default on template for package users, now is the time to do that because its a breaking change but I don't see it as a sensible thing to do tbh. I think defaulting to flat if template is not supplied is sensible behaviour and I see no reason to break backwards compatibility here.

Alternatively, maybe you just mean we should keep the behaviour but push the responsibility for this out a layer? In that case: if we want the npm package and server to have the same behaviour, it probably makes more sense to do it in one place instead of two places but I don't have a really strong opinion on it.

Reply:
Ahh, sorry I should have explained that better.

I think it is a good idea to keep `flat` the default for the library, as it is now. I think this layer is a fine place for that to happen.

The behavior I'm questioning is the library accepting non-existent styles and rendering them as `flat`. I think it should throw an error in that case.

Then the server will need to know which styles are valid, and perform its own validation, to avoid passing a nonexistent style to the renderer.

Reply:
> In that case: if we want the npm package and server to have the same behaviour, it probably makes more sense to do it in one place instead of two places but I don't have a really strong opinion on it.

My suggestion is that the package and the server should reject non-existent styles (e.g. `socail`). That may go against this principle, since it means both the server and the package are doing this validation. I guess my counter-principle is this: it is a good idea for web APIs to validate their inputs early, rather than passing around possible garbage values and catching the errors later. We have a schema for the query params and it's easy to provide a good error message if we reject an invalid value then.

Reply:
yep - that all makes much more sense: 619d850